api/src/main/java/com/example/api/service/activity/result/GraphTaskResultService.java

```diff
@@ -149,13 +156,47 @@ public GraphTaskResult getGraphTaskResult(Long graphTaskId, String email)
         GraphTask graphTask = graphTaskRepo.findGraphTaskById(graphTaskId);
         activityValidator.validateActivityIsNotNull(graphTask, graphTaskId);
 
-        GraphTaskResult result = graphTaskResultRepo.findGraphTaskResultByGraphTaskAndUser(graphTask, user);
-        resultValidator.validateResultIsNotNull(result, graphTaskId, email);
-
-        return result;
+        return getGraphTaskResultWithGraphTaskAndUser(graphTask, user);
     }
 
     public List<GraphTaskResult> getAllGraphTaskResultsForStudent(User student){
         return graphTaskResultRepo.findAllByUser(student);
     }
+
+    public SuperPowerResponse<?> useSuperPower(Long graphTaskId, Long questionId) throws RequestValidationException {
+        User user = getCurrentUserAndValidateStudentAccount();
+        Hero hero = user.getUserHero().getHero();
+
+        GraphTask graphTask = graphTaskRepo.findGraphTaskById(graphTaskId);
+        GraphTaskResult result = getGraphTaskResultWithGraphTaskAndUser(graphTask, user);
+        Question question = questionRepo.findQuestionById(questionId);
+
+        return hero.useSuperPower(heroVisitor, user, result, question);
+    }
+
+    public SuperPowerUsageResponse canSuperPowerBeUsed(Long graphTaskId) throws RequestValidationException {
+        User user = getCurrentUserAndValidateStudentAccount();
+        Hero hero = user.getUserHero().getHero();
+
+        GraphTask graphTask = graphTaskRepo.findGraphTaskById(graphTaskId);
+        GraphTaskResult result = getGraphTaskResultWithGraphTaskAndUser(graphTask, user);
+
+        boolean canBeUsed = hero.canPowerBeUsed(user, result);
+        String message = hero.getCanBeUsedMessage(user, result);
+
+        return new SuperPowerUsageResponse(canBeUsed, message);
+    }
+
+    private GraphTaskResult getGraphTaskResultWithGraphTaskAndUser(GraphTask graphTask, User user) throws EntityNotFoundException {
+        GraphTaskResult result = graphTaskResultRepo.findGraphTaskResultByGraphTaskAndUser(graphTask, user);
+        resultValidator.validateResultIsNotNull(result, graphTask.getId(), user.getEmail());
+        return result;
+    }
+
+    private User getCurrentUserAndValidateStudentAccount() throws WrongUserTypeException {
+        String email = authService.getAuthentication().getName();
+        User user = userRepo.findUserByEmail(email);
+        userValidator.validateStudentAccount(user, email);
+        return user;
+    }
```

Comment:
to chyba można przerzucić do UserService

Reply:
racja